File: Ch09/PerformanceTuning/PerformanceTuning/Predicate.m
//
//  Predicate.m
//  PerformanceTuning
//
//  Created by Rob Warner on 11/5/14.
//  Copyright (c) 2014 Michael Privat and Rob Warner. All rights reserved.
//

#import "Predicate.h"

@implementation Predicate

- (NSString *)runTestWithContext:(NSManagedObjectContext *)context {
  NSMutableString *result = [NSMutableString string];
  
  // Set up the first fetch request
  NSFetchRequest *fetchRequest1 = [NSFetchRequest fetchRequestWithEntityName:@"Selfie"];
  fetchRequest1.predicate = [NSPredicate predicateWithFormat:@"(name LIKE %@) OR (rating < %d)", @"*e*ie*", 5];
  
  // Run the first fetch request and measure
  NSDate *start1 = [NSDate date];
  for (int i = 0; i < 1000; i++) {
    [context reset];
    [context executeFetchRequest:fetchRequest1 error:nil];
  }
  NSDate *end1 = [NSDate date];
  [result appendFormat:@"Slow predicate: %.3f s\n", [end1 timeIntervalSinceDate:start1]];
  
  // Set up the second fetch request
  NSFetchRequest *fetchRequest2 = [NSFetchRequest fetchRequestWithEntityName:@"Selfie"];
  fetchRequest1.predicate = [NSPredicate predicateWithFormat:@"(rating < %d) OR (name LIKE %@)", 5, @"*e*ie*"];
  
  // Run the first fetch request and measure
  NSDate *start2 = [NSDate date];
  for (int i = 0; i < 1000; i++) {
    [context reset];
    [context executeFetchRequest:fetchRequest2 error:nil];
  }
  NSDate *end2 = [NSDate date];
  [result appendFormat:@"Fast predicate: %.3f s\n", [end2 timeIntervalSinceDate:start2]];
  
  return result;
}

@end

File: Ch09/PerformanceTuningSwift/PerformanceTuningSwift/Predicate.swift
//
// From the book Pro iOS Persistence
// Michael Privat and Rob Warner
// Published by Apress, 2014
// Source released under The MIT License
// http://opensource.org/licenses/MIT
//
// Contact information:
// Michael: @michaelprivat -- http://michaelprivat.com -- mprivat@mac.com
// Rob: @hoop33 -- http://grailbox.com -- rwarner@grailbox.com
//

import Foundation
import CoreData

class Predicate : PerfTest {
    func runTestWithContext(context: NSManagedObjectContext) -> String! {
        let result = NSMutableString()
        
        // Set up the first fetch request
        let fetchRequest1 = NSFetchRequest(entityName: "Selfie")
        fetchRequest1.predicate = NSPredicate(format: "(name LIKE %@) OR (rating < %d)", "*e*ie*", 5)
        
        // Run the first fetch request and measure
        let start1 = NSDate()
        for var i=0; i<1000; i++ {
            context.reset()
            context.executeFetchRequest(fetchRequest1, error: nil)
        }
        let end1 = NSDate()
        result.appendFormat("Slow predicate: %.3f s\n", end1.timeIntervalSinceDate(start1))
        
        // Set up the second fetch request
        let fetchRequest2 = NSFetchRequest(entityName: "Selfie")
        fetchRequest2.predicate = NSPredicate(format: "(rating < %d) OR (name LIKE %@)", 5, "*e*ie*")
        
        // Run the first fetch request and measure
        let start2 = NSDate()
        for var i=0; i<1000; i++ {
            context.reset()
            context.executeFetchRequest(fetchRequest2, error: nil)
        }
        let end2 = NSDate()
        result.appendFormat("Fast predicate: %.3f s\n", end2.timeIntervalSinceDate(start2))
        
        return result
    }
}
